### PR TITLE
Fail fast when cache in memory format incompatible with eviction config

### LINKED ISSUE #10716: [cache] Cache eviction not fail fast.
When using the following hz xml only on retrieving the cache, I get a validation error. But the XML file could be validated when loaded. 

```xml
<hazelcast xsi:schemaLocation="http://www.hazelcast.com/schema/config
                               http://www.hazelcast.com/schema/config/hazelcast-config-3.8.xsd"
           xmlns="http://www.hazelcast.com/schema/config"
           xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">

    <group>
        <name>workers</name>
    </group>

    <lite-member enabled="true"/>

    <network>
        <port port-count="200" auto-increment="true">5701</port>
        <join>
            <multicast enabled="false"/>
            <tcp-ip enabled="true">
                <member>192.168.0.1:5701</member>
                <member>192.168.0.2:5701</member>
            </tcp-ip>
        </join>

         <ssl enabled="false"/>
    </network>

    <properties>
        <property name="hazelcast.phone.home.enabled">false</property>
    </properties>

    <license-key>...</license-key>

    <!--MANAGEMENT_CENTER_CONFIG-->

    <native-memory allocator-type="POOLED" enabled="true">
        <size unit="GIGABYTES" value="2" />
        <metadata-space-percentage>20</metadata-space-percentage>
    </native-memory>

    <cache name="cache">
        <eviction size="10000000" max-size-policy="ENTRY_COUNT" eviction-policy="LFU"/>
        <backup-count>1</backup-count>
        <async-backup-count>0</async-backup-count>
        <in-memory-format>NATIVE</in-memory-format>
    </cache>

</hazelcast>
```

```java
ERROR 2017-06-07 19:28:12,284 [Thread-5] com.hazelcast.simulator.worker.testcontainer.TestManager: --------------------------- global prepare of LongStringCacheTest FAILED --------------------------- 
java.lang.IllegalArgumentException: Invalid max-size policy (ENTRY_COUNT) for com.hazelcast.cache.hidensity.impl.nativememory.HiDensityNativeMemoryCacheRecordStore! Only USED_NATIVE_MEMORY_SIZE, USED_NATIVE_MEMORY_PERCENTAGE, FREE_NATIVE_MEMORY_SIZE, FREE_NATIVE_MEMORY_PERCENTAGE are supported.
	at com.hazelcast.cache.hidensity.impl.nativememory.HiDensityNativeMemoryCacheRecordStore.createCacheEvictionChecker(HiDensityNativeMemoryCacheRecordStore.java:121)
	at com.hazelcast.cache.impl.AbstractCacheRecordStore.<init>(AbstractCacheRecordStore.java:150)
	at com.hazelcast.cache.hidensity.impl.nativememory.HiDensityNativeMemoryCacheRecordStore.<init>(HiDensityNativeMemoryCacheRecordStore.java:59)
	at com.hazelcast.cache.EnterpriseCacheService.newNativeRecordStore(EnterpriseCacheService.java:243)
	at com.hazelcast.cache.EnterpriseCacheService.createNewRecordStore(EnterpriseCacheService.java:217)
	at com.hazelcast.cache.impl.CachePartitionSegment.createNew(CachePartitionSegment.java:51)
	at com.hazelcast.cache.impl.CachePartitionSegment.createNew(CachePartitionSegment.java:37)
	at com.hazelcast.util.ConcurrencyUtil.getOrPutSynchronized(ConcurrencyUtil.java:73)
	at com.hazelcast.cache.impl.CachePartitionSegment.getOrCreateRecordStore(CachePartitionSegment.java:67)
	at com.hazelcast.cache.impl.AbstractCacheService.getOrCreateRecordStore(AbstractCacheService.java:278)
```

### PR DESCRIPTION
**Pull request by vbekiaris**
Fixes #10716

### CODE REVIEW

**Comment by ahmetmircik** [comment]
what about programmatically added configurations?

**Comment by vbekiaris** [comment]
Updated to fail on `Cache` proxy creation in case of invalid configuration.

Requires EE counterpart

File: hazelcast/src/main/java/com/hazelcast/cache/impl/CacheProxyUtil.java [hunk 1]
```diff
@@ -236,4 +239,25 @@ public static int getPartitionId(NodeEngine nodeEngine, Data key) {
             }
         }
     }
+
+    public static <K, V> void validateCacheConfig(CacheConfig<K, V> cacheConfig) {
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/cache/impl/CacheProxyUtil.java]: is it possible to combine this `validateCacheConfig` and `checkCacheConfig` methods? seems some duplication exists.

Comment (vbekiaris) [inline hazelcast/src/main/java/com/hazelcast/cache/impl/CacheProxyUtil.java]: Actually one works on `CacheSimpleConfig` and the other on `CacheConfig`. I didn't want to introduce any jcache hierarchy classes in ConfigValidator (in config package) so `checkCacheConfig` works on `CacheSimpleConfig` ( no jcache dependencies) while `validateCacheConfig` implements the same validations on `CacheConfig`.

**Comment by vbekiaris** [comment]
run-lab-run

File: hazelcast-client/src/test/java/com/hazelcast/client/cache/ClientCacheCreationTest.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+import com.hazelcast.core.HazelcastInstance;
+import org.junit.Ignore;
+import org.junit.Test;
+
+import javax.cache.spi.CachingProvider;
+import java.util.Arrays;
+
+public class ClientCacheCreationTest extends CacheCreationTest {
+
+    @Override
```
Comment (Donnerbart) [inline hazelcast-client/src/test/java/com/hazelcast/client/cache/ClientCacheCreationTest.java]: Just some nitpicking. I would prefer this order for readability:
```java
@Test
@Ignore("Only applicable for member-side HazelcastInstance")
@Override
public createInvalidCache_fromDeclarativeConfig_throwsException_fromHazelcastInstanceCreation() {
}
```
You can have `@Override` without the other annotations, so keep it close to the method. And `@Ignore` kind of extends or requires `@Test`, so it goes below it.

File: hazelcast/src/main/java/com/hazelcast/internal/config/ConfigValidator.java [hunk 3]
```diff
@@ -192,6 +193,26 @@ public static void checkEvictionConfig(EvictionPolicy evictionPolicy,
     }
 
     /**
+     * Validates {@link CacheSimpleConfig}.
+     *
+     * @param cacheSimpleConfig
+     */
+    public static void checkCacheConfig(CacheSimpleConfig cacheSimpleConfig) {
+        if (cacheSimpleConfig.getInMemoryFormat() == NATIVE) {
```
[the remark this answers is not in the document]
Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/internal/config/ConfigValidator.java]: Agree with David here, these two are very similar.

**Pull request merged**